src/Optimization/hiopNlpFormulation.hpp

```diff
@@ -94,21 +94,21 @@ class hiopNlpFormulation
    * Wrappers for the interface calls. 
    * Can be overridden for specialized formulations required by the algorithm.
    */
-  virtual bool eval_f(double* x, bool new_x, double& f);
-  virtual bool eval_grad_f(double* x, bool new_x, double* gradf);
+  virtual bool eval_f(hiopVector& x, bool new_x, double& f);
+  virtual bool eval_grad_f(hiopVector& x, bool new_x, double* gradf);
   
-  virtual bool eval_c(double* x, bool new_x, double* c);
-  virtual bool eval_d(double* x, bool new_x, double* d);
-  virtual bool eval_c_d(double* x, bool new_x, double* c, double* d);
+  virtual bool eval_c(hiopVector& x, bool new_x, double* c);
+  virtual bool eval_d(hiopVector& x, bool new_x, double* d);
+  virtual bool eval_c_d(hiopVector& x, bool new_x, double* c, double* d);
   /* the implementation of the next two methods depends both on the interface and on the formulation */
-  virtual bool eval_Jac_c(double* x, bool new_x, hiopMatrix& Jac_c)=0;
-  virtual bool eval_Jac_d(double* x, bool new_x, hiopMatrix& Jac_d)=0;
-  virtual bool eval_Jac_c_d(double* x, bool new_x, hiopMatrix& Jac_c, hiopMatrix& Jac_d);
+  virtual bool eval_Jac_c(hiopVector& x, bool new_x, hiopMatrix& Jac_c)=0;
+  virtual bool eval_Jac_d(hiopVector& x, bool new_x, hiopMatrix& Jac_d)=0;
+  virtual bool eval_Jac_c_d(hiopVector& x, bool new_x, hiopMatrix& Jac_c, hiopMatrix& Jac_d);
 protected:
   //calls specific hiopInterfaceXXX::eval_Jac_cons and deals with specializations of hiopMatrix arguments
-  virtual bool eval_Jac_c_d_interface_impl(double* x, bool new_x, hiopMatrix& Jac_c, hiopMatrix& Jac_d) = 0;
+  virtual bool eval_Jac_c_d_interface_impl(hiopVector& x, bool new_x, hiopMatrix& Jac_c, hiopMatrix& Jac_d) = 0;
 public:
-  virtual bool eval_Hess_Lagr(const double* x, bool new_x, 
+  virtual bool eval_Hess_Lagr(const hiopVector& x, bool new_x, 
 			      const double& obj_factor,  
 			      const double* lambda_eq, 
 			      const double* lambda_ineq, 
```

Comment:
should we use generic hiopVector kernels for duals? From other discussions, eval_Hess_Lagr should be executed on device.

Reply:
`double*` are portable to device from what I understand

Reply:
Actually, @nychiang makes a good point here. If we encapsulate `lambda_eq` and `lambda_ineq` within a HiOp vector then we can eliminate `copyFromStarting` methods that copy from raw arrays and use only the overload that has a HiOp vector as the input.

But, again, this would be beyond the scope of this PR, because `double*` can point to data on device or host, so this is not needed for porting to GPU.

Reply:
I see, that sounds reasonable especially since this is internal / non-interface code.